client/js/widgets/widget.js

```diff
@@ -1027,32 +1027,43 @@ export class Widget extends StateManaged {
           if(jeRoutineLogging)
             jeLoggingRoutineOperationSummary( `elements in '${JSON.stringify(a.in)}'`);
         } else if(a.range) {
+          const rangeProblems = []; // To avoid repeated error messages in each loop iteration
           let range = asArray(a.range);
-          if(range.length == 1)
-            range.unshift(1);
-          if(range.length == 2)
-            range.push(1);
+          
+          if(range.length == 0) {
+            rangeProblems.push(`Empty range given, [1] used.`);
+            range = [1]
+          }
           let start = parseFloat(range[0]);
-          let end = parseFloat(range[1]);
-          let step = parseFloat(range[2]);
           if(isNaN(start)) {
-            problems.push(`Invalid start of range ${JSON.stringify(range[0])}, 1 used`);
+            rangeProblems.push(`Invalid start of range ${JSON.stringify(range[0])}, 1 used`);
             start = 1;
           }
+          
+          if(range.length == 1)
+            range.unshift(1);
+          let end = parseFloat(range[1]);
           if(isNaN(end)) {
-            problems.push(`Invalid end of range ${JSON.stringify(range[1])}, 1 used`);
+            rangeProblems.push(`Invalid end of range ${JSON.stringify(range[1])}, 1 used`);
             end = 1;
           }
+          
+          if(range.length == 2)
+            range.push(end > start ? 1 : -1);
+          let step = parseFloat(range[2]);
           if(isNaN(step) || step == 0) {
-            problems.push(`Invalid step value ${JSON.stringify(range[2])}, 1 used`);
-            step = 1;
+            step = end > start ? 1 : -1;
+            rangeProblems.push(`Invalid step value ${JSON.stringify(range[2])}, ${step} used`);
           }
+          
           if(start>end && step>0 || start<end && step<0) {
             step = -step;
-            problems.push(`Step ${JSON.stringify(range[2])} changed to ${step}`)
+            rangeProblems.push(`Step ${-step} changed to ${step}`)
           }
+          
           for (let index=start; (step > 0) ? index <= end : index >= end; index += step)
             await callWithAdditionalValues({ value: index });
+          problems = rangeProblems;
```

Comment:
Are you sure that you don't dismiss problems here that were added before this block of code?

Reply:
Hmm. I thought problems was re-initialized for each function in a routine, but I see that is not the case. So I could either 1) put things back the way they were and let the error propagate into the loopRoutine executions, or 2) save `problems` before the calls and restore it afterwards. This may be the only place in the module where this is a potential issue. What do you think?

Reply:
It should be re-initialized for each function in a routine. But for example if you use variables inside your `range` property, those could add problems before your block of code. An example would be `The variable ${invalid} is not an object, so indexing it doesn't work.` if you use `range: [ "${invalid.key}" ]`. (I didn't test this.)

Reply:
I'm not sure why your construct (`rangeProblems`) is needed in the first place since you don't add problems in the actual `for` loop. But I didn't look at the code in detail yet and didn't test anything either.

Reply:
The issue is that any error generated by the bounds will be repeated in the problems section of each loop iteration. But it seems too complicated to try to fix properly (the loop iterations could produce their own errors), so I just restored things to the way they were. An extra error message later in the tree than the initial one shouldn't be a big deal.